File: docs/site/samples/typescript/examples/guides_variants_select.ts
/**
 * Selecting a variant at render: omit the name for the default (root body),
 * pass `{ variant }` for that arm. The resolved name comes back on
 * `RenderResult.variant` and the text is that arm's rendered body. Standalone.
 */

import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { Prompt } from "prompting-press";
import { z } from "zod";

// The caller reads the definition; the library does no file I/O itself.
// Resolve the file next to this program (a real app uses its own path).
const defFile = (name: string) => fileURLToPath(new URL(name, import.meta.url));

const SummaryVars = z.object({
  article: z.string().min(1),
  max_words: z.number().int().min(1),
});

const summary = Prompt.fromYaml(readFileSync(defFile("summary.yaml"), "utf8"));
const data = { article: "The Nile floods yearly.", max_words: 20 };

const def = summary.render(SummaryVars, data); // default (root body)
const concise = summary.render(SummaryVars, data, { variant: "concise" });

assert.equal(def.variant, "default");
assert.equal(
  def.text,
  "Summarise the following article in 20 words:\n\nThe Nile floods yearly.",
);
assert.equal(concise.variant, "concise");
assert.equal(concise.text, "In one sentence, summarise: The Nile floods yearly.");

File: docs/site/samples/python/examples/guides_variants_select.py
"""Selecting a variant at render.

Omit the name (or pass ``variant=None``) for the default arm; pass a name for
that arm. The resolved name comes back on ``RenderResult.variant`` and the text
is that arm's rendered body. Standalone program.
"""

from __future__ import annotations

from pathlib import Path

from prompting_press import Prompt
from pydantic import BaseModel, Field

# The caller reads the definition; the library does no file I/O itself.
# Resolve the file next to this program (a real app uses its own path).
_HERE = Path(__file__).parent


class SummaryVars(BaseModel):
    article: str = Field(min_length=1)
    max_words: int = Field(ge=1)


def main() -> None:
    summary = Prompt.from_yaml((_HERE / "summary.yaml").read_text())
    data = {"article": "The Nile floods yearly.", "max_words": 20}

    default = summary.render(SummaryVars, data=data)  # default (root body)
    concise = summary.render(SummaryVars, data=data, variant="concise")

    assert default.variant == "default"
    assert (
        default.text
        == "Summarise the following article in 20 words:\n\nThe Nile floods yearly."
    )
    assert concise.variant == "concise"
    assert concise.text == "In one sentence, summarise: The Nile floods yearly."


if __name__ == "__main__":
    main()
